File: problems/Vertical Order Traversal of a Binary Tree/verticalTraversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<IList<int>> VerticalTraversal(TreeNode root) {
        IDictionary<int, List<(int val, int depth)>> vertLevel2Nodes = new Dictionary<int, List<(int val, int depth)>>();
        Helper(root, 0,0, vertLevel2Nodes);

        if (vertLevel2Nodes.Count == 0)
        {
            return new List<IList<int>>();
        }

        foreach (var vertLevel2Node in vertLevel2Nodes)
        {
            vertLevel2Node.Value.Sort((t1, t2) =>
            {
                var depthCmp =  t1.depth.CompareTo(t2.depth);
                if (depthCmp != 0)
                {
                    return depthCmp;
                }

                return t1.val.CompareTo(t2.val);
            });
        }

        var keys = vertLevel2Nodes.Keys.ToArray();
        Array.Sort(keys);

        IList<IList<int>> res = new List<IList<int>>();

        foreach (var key in keys)
        {
            res.Add(new List<int>());
            foreach (var t in vertLevel2Nodes[key])
            {
                res[res.Count - 1].Add(t.val);
            }
        }

        return res;
    }

    private void Helper(TreeNode node, int key, int depth, IDictionary<int, List<(int val, int depth)>> vertLevel2Nodes)
    {
        if (node == null)
        {
            return;
        }

        if (!vertLevel2Nodes.ContainsKey(key))
        {
            vertLevel2Nodes[key] = new List<(int val, int depth)>();
        }
        vertLevel2Nodes[key].Add((node.val, depth));

        Helper(node.left, key - 1, depth + 1, vertLevel2Nodes);
        Helper(node.right, key + 1, depth + 1, vertLevel2Nodes);
    }
}

File: problems/Vertical Order Traversal of a Binary Tree/verticalTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Queue<NodeAndCol> bfs = new LinkedList<>();
        HashMap<Integer, List<Integer>> store = new HashMap<>();
        PriorityQueue<NodeAndCol> rowHeap = new PriorityQueue<>((a,b)->a.col == b.col ? a.node.val - b.node.val : a.col - b.col);
        int maxCol = Integer.MIN_VALUE;
        int minCol = Integer.MAX_VALUE;

        if (null != root) {
            bfs.add(new NodeAndCol(root, 0));
        }

        while (!bfs.isEmpty()) {
            for (int idx = bfs.size(); 0 < idx; --idx) {
                NodeAndCol nodeAndCol = bfs.poll();
                TreeNode node = nodeAndCol.node;
                int col = nodeAndCol.col;

                if (null != node.left) {
                    bfs.add(new NodeAndCol(node.left, col - 1));
                }

                if (null != node.right) {
                    bfs.add(new NodeAndCol(node.right, 1 + col));
                }

                rowHeap.add(nodeAndCol);
                maxCol = Math.max(maxCol, col);
                minCol = Math.min(minCol, col);
            }

            while (!rowHeap.isEmpty()) {
                NodeAndCol nodeAndCol = rowHeap.poll();
                TreeNode node = nodeAndCol.node;
                int col = nodeAndCol.col;

                if (!store.containsKey(col)) {
                    store.put(col, new ArrayList<>());
                }

                store.get(col).add(node.val);
            }
        }

        for (int col = minCol; maxCol >= col; ++col) {
            if (store.containsKey(col)) {
                result.add(store.get(col));
            }
        }

        return result;
    }

    class NodeAndCol {
        TreeNode node;
        int col;

        public NodeAndCol(TreeNode n, int c) {
            node = n;
            col = c;
        }
    }
}
